File: frontend/src/types/insight.ts
export type ActionType = "watch" | "read" | "test" | "PoC" | "roadmap";
export type Urgency = "critical" | "high" | "medium" | "low";
export type Momentum = "new" | "rising" | "mature";
export type VietnamRelevance = "high" | "medium" | "low";
export type IntelligenceTier = "Tactical" | "Operational" | "Strategic" | "Informational";
export type AdoptionRing = "Adopt" | "Trial" | "Assess" | "Hold";

export interface PracticalIndicators {
  has_code_example?: boolean;
  has_benchmark?: boolean;
  has_api_change?: boolean;
  has_migration_guide?: boolean;
  has_security_patch?: boolean;
}

export interface RecommendationItem {
  action_type: ActionType;
  note: string;
}

export interface InsightListItem {
  id: string;
  title: string;
  summary_short: string | null;
  summary_medium: string | null;
  topics: string[];
  event_type: string | null;
  nature: string | null;
  trust_score: number;
  impact_label: string | null;
  source_url: string;
  confidence: number;
  affected_roles: string[];
  published_at: string | null;
  created_at: string;
  source_id: string;
  source_name: string;
  source_type: string;
  // v2 actionable fields
  signal: string | null;
  why_it_matters: string | null;
  recommendations: Record<string, RecommendationItem> | null;
  risks: string[] | null;
  momentum: Momentum | null;
  urgency: Urgency | null;
  vietnam_relevance: VietnamRelevance | null;
  // v3 taxonomy overhaul
  actionability_score: number | null;
  intelligence_tier: IntelligenceTier | null;
  so_what: string | null;
  adoption_ring: AdoptionRing | null;
  practical_indicators: PracticalIndicators | null;
  primary_image: string | null;
}

export interface InsightReference {
  id: string;
  title: string;
  source_name: string;
  source_url: string;
}

export interface InsightDetail extends InsightListItem {
  summary_medium: string | null;
  status: string;
  updated_at: string;
  source_feed_url: string | null;
  cluster_id: string | null;
  is_primary: boolean;
  references: InsightReference[];
  content_text?: string | null;
}


export interface PaginatedResponse<T> {
  page: number;
  size: number;
  total: number;
  items: T[];
}

File: backend/app/schemas/insight.py
"""Pydantic schemas for Insight API responses."""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict


ActionType = Literal["watch", "read", "test", "PoC", "roadmap"]


class RecommendationItem(BaseModel):
    """A per-role recommendation."""

    action_type: ActionType
    note: str


class InsightReference(BaseModel):
    """A related insight from the same cluster."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    source_name: str
    source_url: str


class InsightListItem(BaseModel):
    """Lightweight insight for list views."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    summary_short: str | None
    summary_medium: str | None
    topics: list[str]
    event_type: str | None
    nature: str | None
    trust_score: float
    impact_label: str | None
    source_url: str
    confidence: float
    affected_roles: list[str]
    published_at: datetime | None
    created_at: datetime
    source_id: uuid.UUID
    source_name: str
    source_type: str
    # v2 actionable fields
    signal: str | None = None
    why_it_matters: str | None = None
    recommendations: dict[str, RecommendationItem] | None = None
    risks: list[str] | None = None
    momentum: str | None = None
    urgency: str | None = None
    vietnam_relevance: str | None = None


class InsightDetail(BaseModel):
    """Full insight detail."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    summary_short: str | None
    summary_medium: str | None
    topics: list[str]
    event_type: str | None
    nature: str | None
    trust_score: float
    impact_label: str | None
    source_url: str
    confidence: float
    status: str
    affected_roles: list[str]
    published_at: datetime | None
    created_at: datetime
    updated_at: datetime
    source_id: uuid.UUID
    source_name: str
    source_type: str
    source_feed_url: str | None
    cluster_id: uuid.UUID | None
    is_primary: bool
    references: list[InsightReference] = []
    # v2 actionable fields
    signal: str | None = None
    why_it_matters: str | None = None
    recommendations: dict[str, RecommendationItem] | None = None
    risks: list[str] | None = None
    momentum: str | None = None
    urgency: str | None = None
    vietnam_relevance: str | None = None
